tex/architecture.tex

```diff
@@ -27,13 +27,38 @@ \section{The Extend Compiler}
+
+  \medskip \noindent This is how the user declares a variable.
+  \begin{lstlisting}
+    [2,2] foo;
+  \end{lstlisting}
+
+  \medskip \noindent This is how the transformer desugars the same code.
+  \begin{lstlisting}
+    rows_of_foo := 2;
+    cols_of_foo := 2;
+    [rows_of_foo, cols_of_foo] foo;
+  \end{lstlisting}
+
+  \medskip \noindent
+  Every expression before or after a comma or colon will become an internal temporary variable in the desugaring process. The transformer also transforms \texttt{\&\&, ||, and switch} into ternary conditionals to enable short-circuiting. Lastly, the transformer performs a small amount of semantic analysis to ensure that there are no duplicate variables within a function, and no duplicate functions within a program.
 
   \subsection{The Semantic Analyzer}
-  The semantic analyzer ensures that Extend functions, variables, expressions, and more are being used properly at compile time, and throws flavorful exceptions to the user so that they may better understand why their program was illegal.
+  The semantic analyzer ensures that Extend functions, variables, expressions, and more are being used properly at compile time, and throws flavorful exceptions to the user so that they may better understand why their program was illegal. In Extend, there are no real type errors, as we attempt to degrade many gracefully. However, the semantic analyzer ensures that functions exist and have the right number of arguments, and that identifiers refer to real variables within their scope.
 
   \subsection{The Code Generator}
-  Provided that the program was deemed legal by \texttt{transform.ml}, the code generator will take the program definition in the abstract syntax tree and generate the appopriate LLVM IR to turn it into a functional program. Instructions to allocate memory, interact with external functions, and platform optimization can be found here.
+  First and foremost, the code generator will validate the program's semantics with \texttt{semant.ml}. It also uses \texttt{transform.ml} to transform the syntax tree as aforementioned, and generate the appopriate LLVM IR to turn it into a functional program. Instructions to allocate memory, interact with external functions, and platform optimization can be found here.
 
   \subsection{The Linker}
   If successful LLVM IR is generated, the linker will adopt the role of building an executable object from the \texttt{.ll} file. This includes compiling it to an object file and linking the runtime environment along with other imported libraries.
+
+  \section{Extend Runtime}
+  % Should talk about runtime.c here
+  When a function is called, we declare a scope, and a set of variable instances. Our LLVM \texttt{main} function is a call to set up variable definitions, then literals, and then a call to the Extend \texttt{main} function. While setting up these variable definitions, we hook up function pointers to the appropriate .. in the array.
```

Comment:
Once a function is called, it recursively looks up what the return value depends on and calculates those values. For this process it is vital that the blueprints created by the Code Generator are correct. The algorithm uses those blueprints to instantiate relevant variables. Ultimately this allows very efficient calculation of the return value without understanding the underlying system.